### PR TITLE
miner: use channels instead of atomics in update loop

### PR DESCRIPTION
**Pull request by MariusVanDerWijden**
This PR changes several different things:
* Adds test cases for the miner loop
* Stops the worker if it wasn't already stopped in `worker.Close()`
* Uses channels instead of atomics in the `miner.update()` loop

### CODE REVIEW

File: miner/miner_test.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+		t.Fatal("miner should be mining")
+	}
+	// Start the downloader
+	mux.Post(downloader.StartEvent{})
+	if miner.Mining() {
+		t.Fatal("miner should not be mining")
+	}
+	// Stop the downloader and wait for the update loop to run
+	mux.Post(downloader.DoneEvent{})
+	time.Sleep(time.Millisecond)
```
Comment (fjl) [inline miner/miner_test.go:95]: This check will fail on slow test machines. There are two options for fixing it:

- Synchronize with the miner for/select loop somehow to ensure the event has been processed
  before checking.
- Add a loop to busy-wait for the state change, e.g. 

  	```
	func waitForMiningState(t *testing.T, m *Miner, mining bool) {
		t.Helper()

		var state bool
		for i := 0; i < 100; i++ {
			if state = m.Mining(); state == mining {
				return
			}
			time.Sleep(10 * time.Millisecond)
		}
		t.Fatal("Mining() == %t, want %t", state, mining)
	}
	```

Comment (MariusVanDerWijden) [reply miner/miner_test.go]: Yeah that makes a lot of sense, thanks!

File: miner/miner.go [hunk 2]
```diff
@@ -96,47 +96,40 @@ func (miner *Miner) update() {
...
-					miner.Start(miner.coinbase)
+					miner.SetEtherbase(miner.coinbase)
+					miner.worker.start()
 				}
-				// stop immediately and ignore all further pending events
-				return
 			}
+		case addr := <-miner.startCh:
+			miner.SetEtherbase(addr)
+			miner.worker.start()
```
Comment (holiman) [inline miner/miner.go:114]: shouldn't this set `shouldStart=true`

File: miner/miner.go [hunk 3]
```diff
@@ -96,47 +96,40 @@ func (miner *Miner) update() {
...
+		case <-miner.stopCh:
+			miner.worker.stop()
```
Comment (holiman) [inline miner/miner.go:116]: ... and shouldn't this set `shouldStart=false` ? 
In general. I'd prefer `miningRequested` over `shouldStart`

**Review by holiman — COMMENTED** [review]
I think this was merged a bit prematurely...?

**Comment by fjl** [comment]
I think it's not necessary to set `shouldStart = true` in the start clause because the logic that handles downloader events simply checks whether mining was enabled when sync starts. It will then take care to re-start it when the sync finishes.

However, I think there is a bug in the stop logic: it will also restart mining when Stop() was called during sync, because the select case for stop doesn't set `shouldStart`.

**Pull request merged**